src/main.c

```diff
@@ -2011,6 +2011,23 @@ static gboolean service_start(int argc, char **argv)
 		return TRUE;
 	}
 
+	/* Boot ID-based boot vs service restart detection */
+	if (g_strcmp0(r_context()->global_state->boot_id, r_context()->boot_id) == 0) {
+		g_message("Restarted RAUC service");
+	} else {
+		RaucSlot *booted_slot = r_slot_get_booted(r_context()->config->slots);
+		g_message("Booted into %s (%s)", booted_slot->name, booted_slot->bootname);
+
+		/* update boot ID */
+		g_free(r_context()->global_state->boot_id);
+		r_context()->global_state->boot_id = g_strdup(r_context()->boot_id);
+
+		if (!r_global_state_save(r_context()->statepath, r_context()->global_state, &ierror)) {
+			g_warning("Failed saving global state");
+			g_clear_error(&ierror);
+		}
```

Comment:
I find passing data from `r_context()` into a different module strange. Perhaps this should a helper in `context.h` or `r_global_state_save` should access the context directly?

Reply:
We discussed that (unlike for the load function which gets called before having context set up) passing path and state object explicitly is not required here. I will check if we can also use a wrapper method in `context.c` to hide the direct call to slot saving method.

Reply:
You decided against the wrapper, but with out the explicit parameters, I'm fine with this.